core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1080,6 +1092,9 @@ class Partition(val topicPartition: TopicPartition,
     // decide whether to only fetch from leader
     val localLog = localLogWithEpochOrException(currentLeaderEpoch, fetchOnlyFromLeader)
 
+    // Check that the partition leader is recovering from an unclean leader election.
+    validateLeaderRecoveryState()
```

Comment:
I didn't implement this now because only the leader can accept PRODUCE requests and the leader recovers immediately. In this implementation only the follower needs to wait for the leader to recover so it can accept FETCH requests.